add copy-publishers-from private subcommand and use it

lotheac opened the pull request:
This fixes #2 and greatly simplifies the ipkg zone creation and attach scripts. I tested that both actions still work:

    hipster ~/pkg5 % pfexec zoneadm -z foo install
    A ZFS file system has been created for this zone.
    Sanity Check: Looking for 'entire' incorporation.
           Image: Preparing at /zones/foo/root.

       Publisher: Using omnios (http://pkg.omniti.com/omnios/bloody/).
       Publisher: Using on-nightly (file:///ws/illumos-omnios/packages/i386/nightly/repo.redist/).
       Publisher: Using niksula.hut.fi (http://pkg.niksula.hut.fi/).
           Cache: Using /var/pkg/publisher.
      Installing: Packages (output follows)
    Packages to install: 386
    Mediators to change:   2
     Services to change:   4

    DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
    Completed                            386/386   36374/36374  324.0/324.0  753k/s

    PHASE                                          ITEMS
    Installing new actions                   59619/59619
    Updating package state database                 Done 
    Updating package cache                           0/0 
    Updating image state                            Done 
    Creating fast lookup database                   Done 

            Note: Man pages can be obtained by installing pkg:/system/manual
     Postinstall: Copying SMF seed repository ... done.
            Done: Installation completed in 466.477 seconds.

      Next Steps: Boot the zone, then log into the zone console (zlogin -C)
                  to complete the configuration process.
    hipster ~/pkg5 % pfexec pkg -R /zones/foo/root list -v -af entire
    FMRI                                                                         IFO
    pkg://omnios/entire@11-0.151017:20151211T141541Z                             i--
    pkg://omnios/entire@11-0.151017:20151211T033227Z                             ---
    pkg://omnios/entire@11-0.151017:20151106T034315Z                             ---
    hipster ~/pkg5 % pfexec zoneadm -z foo detach
    hipster ~/pkg5 % pfexec pkg -R /zones/foo/root update entire@11-0.151017:20151106T034315Z 
    Packages to update: 1

    PHASE                                          ITEMS
    Removing old actions                             1/1
    Installing new actions                           1/1
    Updating modified actions                        1/1
    Updating package state database                 Done 
    Updating package cache                           1/1 
    Updating image state                            Done 
    Creating fast lookup database                   Done 
    hipster ~/pkg5 % pfexec zoneadm -z foo attach -u
    Log File: /var/tmp/foo.attach_log.50aW5L
                   Attach Path: /zones/foo/root
            Attach ZFS Dataset: rpool/zones/foo/ROOT/zbe

                    Installing: Using pre-existing data in zonepath
           Global zone version: entire@11-0.151017:20151211T141541Z
       Non-Global zone version: entire@11-0.151017:20151106T034315Z
                         Cache: Using /var/pkg/publisher.
      Updating non-global zone: Output follows
    Packages to update: 1

    PHASE                                          ITEMS
    Removing old actions                             1/1
    Installing new actions                           1/1
    Updating modified actions                        1/1
    Updating package state database                 Done 
    Updating package cache                           1/1 
    Updating image state                            Done 
    Creating fast lookup database                   Done 

      Updating non-global zone: Zone updated.
                        Result: Attach Succeeded.

and that the signature-policy and approved CAs are actually copied:

    hipster ~/pkg5 % pfexec pkg -R /zones/foo/root publisher niksula.hut.fi

                Publisher: niksula.hut.fi
                    Alias: 
               Origin URI: http://pkg.niksula.hut.fi/
                  SSL Key: None
                 SSL Cert: None
              Client UUID: 2c8248f2-afc5-11e5-b8fe-82082085168c
          Catalog Updated: Tue Dec 22 15:55:43 2015
             Approved CAs: 5b14cb37b522013079016a6867c65e491a57efb3
                  Enabled: Yes
    hipster ~/pkg5 % pfexec pkg -R /zones/foo/root publisher omnios

                Publisher: omnios
                    Alias: 
               Origin URI: http://pkg.omniti.com/omnios/bloody/
                  SSL Key: None
                 SSL Cert: None
              Client UUID: eac58714-5b26-11e5-9665-802590aba566
          Catalog Updated: Fri Dec 11 14:18:34 2015
                  Enabled: Yes
               Properties:
                           signature-policy = verify

lotheac commented:
One thing I notice now is that the copied publishers have the same Client UUID as the source publishers. I don't know if that actually matters for any OmniOS use case but it's not difficult to regenerate UUIDs on copy if you think that should be done. The documentation states it's used to identify the client to networked publishers (which I actually think is a slightly shady thing to do in the first place): from pkg(1):

```
   send-uuid

       (boolean) Send the image's Universally Unique Identifier (UUID)
       when performing network operations. Although users can disable this
       option, some network repositories might refuse to talk to clients
       that do not supply a UUID.

       Default value: True
```

danmcd commented:
A change this big is going to need some additional testing:

Effect these changes on, say, r151016, and then upgrade the whole machine to bloody (NOTE: you'll have to disable signature policies, or start with r151014 and upgrade to r151016).  Do it both ways (using zoneadm detach, and using the lipkg/change-publishers method).  I expect you still have to change the NGZs' publishers for "omnios", but please make sure that operation works.

lotheac commented:
I amended the changeset to work with GZ publishers that have multiple origins
(ran into that while testing), and tested as follows.

On Mon, Jan 04 2016 05:26:27 -0800, Dan McDonald wrote:

> A change this big is going to need some additional testing:
> 
> 1.) Effect these changes on, say, r151016, and then upgrade the whole machine to bloody.  Do it both ways (using zoneadm detach, and using the lipkg/change-publishers method).  I expect you still have to change the NGZs' publishers for "omnios", but please make sure that operation works.

First, I created and installed three zones (two ipkg and one lipkg brand) using
the published 151016 pkg bits.

NB: Because there are higher versioned libxml2 and dns/bind packages available
in 151016 than in bloody, I had to downgrade them manually at this point (on
all zones; zone installation does not constrain them to GZ version) to be able
to upgrade cleanly later.

I detached one of the ipkg zones to prepare for the upgrade:

```
  ID NAME             STATUS     PATH                           BRAND    IP    
   0 global           running    /                              ipkg     shared
   3 ipkg2            running    /zones/ipkg2                   ipkg     shared
   4 lipkg            running    /zones/lipkg                   lipkg    shared
   - ipkg1            configured /zones/ipkg1                   ipkg     shared
```

At this point, I changed the global zone's omnios publisher to point to bloody,
as well as added my own origin for the omnios'publisher (which contains the
updated pkg bits):

```
hipster ~ % pfexec pkg set-publisher -G http://pkg.omniti.com/omnios/r151016/ -g http://pkg.omniti.com/omnios/bloody --set-property signature-policy=verify omnios
hipster ~ % pfexec pkg set-publisher -g ~/newpkg omnios
```

This was also necessary to do for the lipkg zone (repeat the above with -R
/zones/lipkg/root), or upgrade would fail.

pkg complains about being out of date when attempting to upgrade now, but
that's ok, let's do what it asks.

```
hipster ~ % pfexec pkg update --be-name=omnios-bloody
WARNING: pkg(5) appears to be out of date, and should be updated before
running update.  Please update pkg(5) by executing 'pkg install
pkg:/package/pkg' as a privileged user and then retry the update.
hipster ~ % pfexec pkg install pkg:/package/pkg    
            Packages to update:  2
       Create boot environment: No
Create backup boot environment: No

Planning linked: 0/1 done; 1 working: zone:lipkg
Linked image 'zone:lipkg' output:
| Packages to update: 2
`
Planning linked: 1/1 done
DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                                2/2         21/21      0.4/0.4    0B/s

Downloading linked: 0/1 done; 1 working: zone:lipkg
Downloading linked: 1/1 done
PHASE                                          ITEMS
Removing old actions                             4/4
Installing new actions                           2/2
Updating modified actions                      26/26
Updating package state database                 Done 
Updating package cache                           2/2 
Updating image state                            Done 
Creating fast lookup database                   Done 
Reading search index                            Done 
Updating search index                            2/2 
Executing linked: 0/1 done; 1 working: zone:lipkg
Executing linked: 1/1 done
```

NB: now we have the new pkg bits running in the 151016 environment, so we might
as well try creating another set of three zones at this point to see if those
behave differently. But to ensure we get the same results, we must first revert
the publisher changes (because new zones use the GZ's publishers).

```
hipster ~ % pfexec pkg set-publisher -g http://pkg.omniti.com/omnios/r151016/ -G http://pkg.omniti.com/omnios/bloody/ --set-property signature-policy=require-signatures omnios        
hipster ~ % pfexec pkg set-publisher -G ~/newpkg omnios
hipster ~ % for z in ipkg1 ipkg2 lipkg; do zonecfg -z $z export|sed 's,zonepath=/zones/,zonepath=/zones/new_,' | pfexec zonecfg -z new_$z && pfexec zoneadm -z new_$z install; done
new_ipkg1: No such zone configured
Use 'create' to begin configuring a new zone.
A ZFS file system has been created for this zone.
Sanity Check: Looking for 'entire' incorporation.
       Image: Preparing at /zones/new_ipkg1/root.

   Publisher: Using omnios (http://pkg.omniti.com/omnios/r151016/).
   Publisher: Using niksula.hut.fi (http://pkg.niksula.hut.fi/).
       Cache: Using /var/pkg/publisher.
  Installing: Packages (output follows)
Packages to install: 386
Mediators to change:   1
 Services to change:   4

DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                            386/386   36744/36744  324.2/324.2  750k/s

PHASE                                          ITEMS
Installing new actions                   59652/59652
Updating package state database                 Done 
Updating package cache                           0/0 
Updating image state                            Done 
Creating fast lookup database                   Done 

        Note: Man pages can be obtained by installing pkg:/system/manual
 Postinstall: Copying SMF seed repository ... done.
        Done: Installation completed in 461.009 seconds.

  Next Steps: Boot the zone, then log into the zone console (zlogin -C)
              to complete the configuration process.

[similar output for two other zones omitted]
```

So, that worked fine. Next, I detached new_ipkg1 and downgraded the libxml2 and
dns/bind packages in the new zones to match the zones we created with the old
bits. The situation is now as follows, with the "new_" zones having been
created with the updated pkg bits:

```
  ID NAME             STATUS     PATH                           BRAND    IP    
   0 global           running    /                              ipkg     shared
   3 ipkg2            running    /zones/ipkg2                   ipkg     shared
   4 lipkg            running    /zones/lipkg                   lipkg    shared
   5 new_ipkg2        running    /zones/new_ipkg2               ipkg     shared
   6 new_lipkg        running    /zones/new_lipkg               lipkg    shared
   - ipkg1            configured /zones/ipkg1                   ipkg     shared
   - new_ipkg1        configured /zones/new_ipkg1               ipkg     shared
```

At this point, we are ready to actually do the upgrade. So change the publisher
to bloody again and add the local origin for the new pkg (taking care to do the
same for new_lipkg; the old lipkg zone already has the bloody origin because we
didn't revert it above):

```
hipster ~ % pfexec pkg set-publisher -G http://pkg.omniti.com/omnios/r151016/ -g http://pkg.omniti.com/omnios/bloody/ -g ~/newpkg/ --set-property signature-policy=verify omnios
hipster ~ % pfexec pkg -R /zones/new_lipkg/root set-publisher -G http://pkg.omniti.com/omnios/r151016/ -g http://pkg.omniti.com/omnios/bloody/ -g ~/newpkg/ --set-property signature-policy=verify omnios
```

And the actual upgrade goes fine:

```
pfexec pkg update -vn  62.15s user 7.36s system 37% cpu 3:05.44 total
hipster ~ % pfexec pkg update --be-name=omnios-bloody   
            Packages to remove:   1
            Packages to update: 427
       Create boot environment: Yes
Create backup boot environment:  No

Planning linked: 0/2 done; 1 working: zone:lipkg
Linked image 'zone:lipkg' output:
|  Packages to remove:   1
| Packages to install:   1
|  Packages to update: 383
| Mediators to change:   1
|  Services to change:   2
`
Planning linked: 1/2 done; 1 working: zone:new_lipkg
Linked image 'zone:new_lipkg' output:
|  Packages to remove:   1
| Packages to install:   1
|  Packages to update: 385
| Mediators to change:   1
|  Services to change:   2
`
Planning linked: 2/2 done
DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                            428/428     9194/9194  489.9/489.9  1.5M/s

Downloading linked: 0/2 done; 1 working: zone:lipkg
Downloading linked: 1/2 done; 1 working: zone:new_lipkg
Downloading linked: 2/2 done
PHASE                                          ITEMS
Removing old actions                       2563/2563
Installing new actions                     2468/2468
Updating modified actions                12770/12770
Updating package state database                 Done 
Updating package cache                       428/428 
Updating image state                            Done 
Creating fast lookup database                   Done 
Reading search index                            Done 
Building new search index                    429/429 
Executing linked: 0/2 done; 1 working: zone:lipkg
Executing linked: 1/2 done; 1 working: zone:new_lipkg
Executing linked: 2/2 done

A clone of r151016 exists and has been updated and activated.
On the next boot the Boot Environment omnios-bloody will be
mounted on '/'.  Reboot when ready to switch to this updated BE.


---------------------------------------------------------------------------
NOTE: Please review release notes posted at:

http://omnios.omniti.com/ReleaseNotes
---------------------------------------------------------------------------

pfexec pkg update --be-name=omnios-bloody  248.80s user 55.07s system 28% cpu 17:37.19 total
```

After rebooting we find that our lipkg zones are already upgraded as they
should be. We also have the detached zones ipkg1 and new_ipkg1 which need to be
attached with -u; that works fine:

```
hipster ~ % pfexec zoneadm -z ipkg1 attach -u
Log File: /var/tmp/ipkg1.attach_log.sPai9f
               Attach Path: /zones/ipkg1/root
        Attach ZFS Dataset: rpool/zones/ipkg1/ROOT/zbe

                Installing: Using pre-existing data in zonepath
       Global zone version: entire@11-0.151017:20151211T141541Z
   Non-Global zone version: entire@11-0.151016:20151202T161203Z
                     Cache: Using /var/pkg/publisher.
  Updating non-global zone: Output follows
 Packages to remove:   1
Packages to install:   1
 Packages to update: 384
Mediators to change:   1
 Services to change:   2

DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                            386/386     7461/7461  183.3/183.3  144k/s

PHASE                                          ITEMS
Removing old actions                       2271/2271
Installing new actions                     2248/2248
Updating modified actions                10286/10286
Updating package state database                 Done 
Updating package cache                       385/385 
Updating image state                            Done 
Creating fast lookup database                   Done 

  Updating non-global zone: Zone updated.
                    Result: Attach Succeeded.
pfexec zoneadm -z ipkg1 attach -u  56.18s user 12.53s system 52% cpu 2:11.99 total
```

Additionally, we can look at the zones that were not updated and see
that the approved CA has been copied for the niksula.hut.fi publisher
only for the zone created by the new bits (demonstrating that this does
solve #2):

```
hipster ~ % for z in ipkg2 new_ipkg2; do pfexec pkg -R /zones/$z/root publisher niksula.hut.fi; done

            Publisher: niksula.hut.fi
                Alias: 
           Origin URI: http://pkg.niksula.hut.fi/
              SSL Key: None
             SSL Cert: None
          Client UUID: 25049a7c-b32a-11e5-aceb-82082085168c
      Catalog Updated: December 22, 2015 03:55:43 PM 
              Enabled: Yes

            Publisher: niksula.hut.fi
                Alias: 
           Origin URI: http://pkg.niksula.hut.fi/
              SSL Key: None
             SSL Cert: None
          Client UUID: 2c8248f2-afc5-11e5-b8fe-82082085168c
      Catalog Updated: December 22, 2015 03:55:43 PM 
         Approved CAs: 5b14cb37b522013079016a6867c65e491a57efb3
              Enabled: Yes
```

I think all that sufficiently demonstrates that the new bits still work
and that they solve new problems.

## 

Lauri Tirkkonen | lotheac @ IRCnet

danmcd commented:
Wow, that's nice & thorough.  I will pull this in no later than COB tomorrow, US/Eastern.  I need to locally update my bloody build box with some other things prior to accepting this and spinning the world again.  This WILL, however, make the next bloody update that hits the public repo servers.

danmcd commented:
Actually, I think I'll pull & build now.
